assembly/src/ast/visit.rs

```diff
@@ -0,0 +1,836 @@
+use core::ops::ControlFlow;
+
+use crate::{ast::*, Felt, Span};
+
+pub trait Visit<T = ()> {
```

Comment:
It's purely ergonomics/syntax sugar - the vast majority of visitors will use `()` as `T` here, the exception being visitors which want to support early return (i.e. `ControlFlow::Break`) with a payload of some kind. We use that in two impls that I know of, but the others all use `()`, and in those cases we don't have to mention `()` at all.

This trick is used quite frequently in many places in the standard library. All of the arithmetic traits, e.g. `Add`, things like `PartialEq`, etc., default `T` to `Self` - or whatever is appropriate in a given situation, but it is frequently `Self`. For example `HashMap` defaults its `BuildHasher` generic argument to a concrete value, so that most users only need to specify `HashMap<K, V>` when referring to it.

Anyway, it's not needed, but it does make things a little nicer to work with.